libp2p/network/network.go

```diff
@@ -0,0 +1,219 @@
+	addrBook, err := addrBookProvider.GetStakedAddrBookAtHeight(currentHeightProvider.CurrentHeight())
+	if err != nil {
+		log.Fatal().Err(err).Msg("getting staked address book")
+	}
+
+	addrBookMap := make(types.AddrBookMap)
+	for _, poktPeer := range addrBook {
+		addrBookMap[poktPeer.Address.String()] = poktPeer
+		pubKey, err := identity.PubKeyFromPoktPeer(poktPeer)
+		if err != nil {
+			return nil, ErrNetwork(fmt.Sprintf(
+				"converting peer public key, pokt address: %s", poktPeer.Address,
+			), err)
+		}
+		peer, err := identity.PeerAddrInfoFromPoktPeer(poktPeer)
+		if err != nil {
+			return nil, ErrNetwork(fmt.Sprintf(
+				"converting peer info, pokt address: %s", poktPeer.Address,
+			), err)
+		}
+
+		host_.Peerstore().AddAddrs(peer.ID, peer.Addrs, DefaultPeerTTL)
+		if err := host_.Peerstore().AddPubKey(peer.ID, pubKey); err != nil {
+			return nil, ErrNetwork(fmt.Sprintf(
+				"adding peer public key, pokt address: %s", poktPeer.Address,
+			), err)
+		}
+	}
+
+	return &libp2pNetwork{
+		logger: log,
+		// TODO: is it unconventional to set bus here?
+		bus:         bus,
+		host:        host_,
+		topic:       topic,
+		addrBookMap: addrBookMap,
+	}, nil
+}
+
+// NetworkBroadcast uses the configured pubsub router to broadcast data to peers.
+func (p2pNet *libp2pNetwork) NetworkBroadcast(data []byte) error {
+	// IMPROVE: receive context in interface methods?
+	ctx := context.Background()
+
+	// NB: Routed send using pubsub
+	if err := p2pNet.topic.Publish(ctx, data); err != nil {
+		return ErrNetwork("unable to publish to topic", err)
+	}
+	return nil
+}
+
+// NetworkSend connects sends data directly to the specified peer.
+func (p2pNet *libp2pNetwork) NetworkSend(data []byte, poktAddr poktCrypto.Address) error {
+	// IMPROVE: add context to interface methods.
+	ctx := context.Background()
+
+	selfPoktAddr, err := p2pNet.GetBus().GetP2PModule().GetAddress()
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"sending to poktAddr: %s", poktAddr,
+		), err)
+	}
+
+	// NB: don't send to self.
+	if selfPoktAddr.Equals(poktAddr) {
+		return nil
+	}
+
+	poktPeer, ok := p2pNet.addrBookMap[poktAddr.String()]
+	if !ok {
+		// NB: this should not happen.
+		return ErrNetwork("", fmt.Errorf(
+			"peer not found in address book, pokt address: %s", poktAddr,
+		))
+	}
+
+	peerAddrInfo, err := identity.PeerAddrInfoFromPoktPeer(poktPeer)
+	if err != nil {
+		return ErrNetwork("parsing peer multiaddr", err)
+	}
+
+	stream, err := p2pNet.host.NewStream(ctx, peerAddrInfo.ID, protocol.PoktProtocolID)
+	if err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"opening peer stream, pokt address: %s", poktAddr,
+		), err)
+	}
+
+	if _, err := stream.Write(data); err != nil {
+		return ErrNetwork(fmt.Sprintf(
+			"writing to stream (peer address: %s)", poktAddr,
+		), err)
+	}
+	defer func() {
+		if err := stream.Close(); err != nil {
+			p2pNet.logger.Error().Err(err)
+		}
+	}()
+
+	// TECHDEBT: check if there's a more conventional way to send
```

Comment:
Hmmm.. So topics let us publish to everyone (based on the configuration) but 1:1 comms require a stream?

Reply:
Indeed. `pubsub` is a higher-level concept built on top of the hosts' streams as far as I understand.

Reply:
We could surely conceive of a way to use pubsub for 1:1 if we wanted to but I think it would be over-engineered which could make debugging more difficult. Depending on the properties that we want a "1:1 message exchange" to have, we may have to do additional work in the implementation but maybe also at the protocol level to ensure them. For example, in the most basic implementation, anyone may be able to learn the topic used for a "1:1 exchange". On the other hand, if we had a good reason to consider that route, it may worth taking a quick look to see if libp2p supports some kind of access control around topics or something. It does include options for message signing and verification so it's not too much of a stretch to imagine.

Reply:
I know you were looking at some other codebases, so I think I'd be interested in learning what other chains using libp2p are using.